sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -264,12 +281,12 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
         df,
         sink,
         outputMode,
-        useTempCheckpointLocation = true,
+        useTempCheckpointLocation = isTempCheckpointLocationAvailable,
         trigger = trigger)
     } else {
       val (useTempCheckpointLocation, recoverFromCheckpointLocation) =
         if (source == "console") {
-          (true, false)
+          (isTempCheckpointLocationAvailable, false)
```

Comment:
@jerryshao whether `useTempCheckpointLocation` is `true` or `false` is still based on the type of Sink, but there is an additional constraint: the `defaultFs` must be the same as the `tmpFs`. This is the reason of this patch. Indeed, otherwise the metadata whose path is based on the checkpoint directory will be tried to be created on `defaultFs`.
If it is different from the `tmpFs`, you get a `PermissionDenied` exception which is actually caused by the fact that the temp dir is created on the `tmpFs` and not on the `defaultFs`. With this PR, instead, you will get an `AnalysisException` telling you to set a `checkpointLocation`, which is a much more meaningful exception, easy to understand and to fix.